crates/edr_solidity/src/contracts_identifier.rs

```diff
@@ -0,0 +1,703 @@
+        // We check if we got to match the entire executable bytecode, and are just
+        // stuck because of the metadata. If that's the case, we can assume that
+        // any descendant will be a valid Bytecode, so we just choose the most
+        // recently added one.
+        //
+        // The reason this works is that there's no chance that Solidity includes an
+        // entire bytecode (i.e. with metadata), as a prefix of another one.
+
+        if is_matching_metadata(code, matched_bytes) {
+            let last_suffix = self.tree.root().descendant_suffixes().last();
+
+            if let Some(last_suffix) = last_suffix {
+                // TODO: this should be the last one in chronological insertion order
+                let mut descendant = code.to_vec()[..node.bytes_matched_before()].to_vec();
+                descendant.extend(last_suffix);
+
+                return self.bytecodes.get(&calculate_hash(&descendant)).copied();
+            }
+        }
+
+        None
+    }
+}
+
+fn zero_out_addresses(code: &Bytes, offsets: &[usize]) -> Bytes {
+    let references = offsets
+        .iter()
+        .map(|offset| ImmutableReference {
+            offset: *offset,
+            length: 20,
+        })
+        .collect::<Vec<_>>();
+
+    zero_out_slices(code, &references)
+}
+
+fn zero_out_slices(code: &Bytes, immutable_references: &[ImmutableReference]) -> Bytes {
+    let mut result = code.to_vec();
+
+    for reference in immutable_references {
+        for byte in result
+            .iter_mut()
+            .skip(reference.offset)
+            .take(reference.length)
+        {
+            *byte = 0;
+        }
+    }
+
+    result.into()
+}
+
+fn is_create_trace(trace: &EvmMessageTrace) -> bool {
+    match trace {
+        EvmMessageTrace::Create(_) => true,
+        EvmMessageTrace::Call(_) => false,
+    }
+}
+
+fn normalize_library_runtime_bytecode_if_necessary(bytecode: Bytes) -> Bytes {
+    // Libraries' protection normalization:
+    // Solidity 0.4.20 introduced a protection to prevent libraries from being
+    // called directly. This is done by modifying the code on deployment, and
+    // hard-coding the contract address. The first instruction is a PUSH20 of
+    // the address, which we zero-out as a way of normalizing it. Note that it's
+    // also zeroed-out in the compiler output.
+
+    if let Some(first_opcode) = bytecode.first() {
+        if *first_opcode == opcode::PUSH20 {
+            return zero_out_addresses(&bytecode, &[1]);
+        }
+    }
+
+    bytecode
+}
+
+fn calculate_hash<T: Hash>(t: &T) -> u64 {
+    let mut s = DefaultHasher::new();
+    t.hash(&mut s);
+    s.finish()
+}
+
+fn is_matching_metadata(code: &[u8], last_byte: usize) -> bool {
+    let mut byte = 0;
+    while byte < last_byte {
+        let opcode = code[byte];
+
+        // Solidity always emits REVERT INVALID right before the metadata
+        if opcode == opcode::REVERT && code[byte + 1] == opcode::INVALID {
```

Comment:
Indexing using `[..]` can panic and should thus be avoided.

You can rewrite this using https://doc.rust-lang.org/std/primitive.slice.html#method.windows

Reply:
Thanks for this, it was a dumb oversight on my part (the same code is "safe" in javascript, with big quotes). I tried to add a test that would crash because of this but that made me realize that right now it's actually impossible. But proving that (and making sure that it never becomes possible) is way harder than just avoiding indexing with `[]` :sweat_smile:

That being said, I'm not sure this method can easily be re-written with `.windows` because I also need to advance the window by an arbitrary number (explanation of why this is necessary below). What do you think about this?

```rust
fn is_matching_metadata(code: &[u8]) -> bool {
    let mut byte = 0;
    while let (Some(opcode), Some(next_opcode)) = (code.get(byte), code.get(byte + 1)) {
        if *opcode == opcode::REVERT && *next_opcode == opcode::INVALID {
            return true;
        }

        byte += opcode_length(*opcode);
    }

    false
}
```

and the call becomes `is_matching_metadata(&code[..matched_bytes])`.

---

The reason we use `opcode_length` to skip forward is not (just) an optimization. You could have, for example, a `PUSH20` whose argument, by chance, contains the `REVERT INVALID` sequence.

Reply:
That would work. Alternatively, you can use `.get(..).expect("Some reason")` to signal that you'd never expect this to fail and why.

That's also valid and safe. It will result in a reason for the panic at runtime, but at least it's explicitly documented and can be tested for

Reply:
Refactored in 2af60e1d241973a23911fcd5201b6f72d0c88e7a